src/libhirte/common/config.c

```diff
@@ -0,0 +1,259 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+
+#include <asm-generic/errno-base.h>
+#include <assert.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "libhirte/hashmap/hashmap.h"
+#include "libhirte/ini/ini.h"
+
+#include "common.h"
+#include "config.h"
+
+/*
+ * Structure holding the configuration is just hiding that it uses a hashmap structure internally
+ */
+struct config {
+        struct hashmap *cfg_store;
+        char *default_section;
+};
+
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+static int option_compare(const void *a, const void *b, void *udata) { // NOLINT
+        const struct config_option *option_a = a;
+        const struct config_option *option_b = b;
+        return !(streq(option_a->section, option_b->section) && streq(option_a->name, option_b->name));
+}
+#pragma GCC diagnostic pop
+
+static uint64_t option_hash(const void *item, uint64_t seed0, uint64_t seed1) {
+        const struct config_option *option_a = item;
+        char *sect_and_name = NULL;
+        // there is no way how to tell hashmap implementation that there was an error during hash generation
+        assert(asprintf(&sect_and_name, "%s=%s", option_a->section, option_a->name) != -1);
+        uint64_t hash = hashmap_sip(sect_and_name, strlen(sect_and_name), seed0, seed1);
+        free(sect_and_name);
+        return hash;
+}
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+/*
+ * Handle used to store option parsed from a file into a configuration.
+ *
+ * Returns 1 on success, 0 on error as per ini library requirements.
+ */
+static int parsing_handler(void *user, const char *section, const char *name, const char *value) { // NOLINT
+        if (user == NULL) {
+                return 0;
+        }
+        struct config *config = (struct config *) user;
+        if (cfg_s_set_value(config, section, name, value) != 0) {
+                return 0;
+        }
+        return 1;
+}
+#pragma GCC diagnostic pop
+
+int cfg_initialize(struct config **config) {
+        struct config *new_cfg = malloc(sizeof(struct config));
+        if (new_cfg == NULL) {
+                // error during configuration structure initialization
+                return -ENOMEM;
+        }
+        new_cfg->cfg_store = hashmap_new(
+                        sizeof(struct config_option), 0, 0, 0, option_hash, option_compare, NULL, NULL);
+        if (new_cfg->cfg_store == NULL) {
+                // error during hashmap initialization
+                free(new_cfg);
+                return -ENOMEM;
+        }
+        char *section_copy = strdup(CFG_SECT_GLOBAL);
+        if (section_copy == NULL) {
+                // error during default section initialization
+                free(new_cfg->cfg_store);
```

Comment:
should be hashmap_free(), not free(). Although, optionally you can call cfg_dispose() instead, if you make sure cfg_dispose handles some of the fields being NULL.